kv.go

```diff
@@ -10,83 +10,159 @@ import "context"
 // different behavior.
 var DefaultIterOptions = IterOptions{}
 
+// IterOptions contains the full set of available iterator options,
+// it can be provided when creating an [Iterator] from a [Store].
 type IterOptions struct {
-	// Prefix iteration, all keys *must* contain
-	// the designated prefix. Prefix takes precendent
-	// over range (start/end). If Prefix is nil, range
-	// will be used (regardless if start/end is nil)
+	// Prefix iteration, only keys beginning with the designated prefix
+	// with the given prefix will be yielded.
+	//
+	// Keys exactly matching the provided `Prefix` value will not be
+	// yielded.
+	//
+	// Providing a Prefix value should cause the Start and End options
+	// to be ignored, although this is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/35
 	Prefix []byte
 
-	// Range iteration from start (inclusive) to end explusive.
-	// Start must be lexigraphically less
-	// then end, unless `nil` is used.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically greater than or
+	// equal to this value.
 	//
-	// Start defined as nil will iterate from the first
-	// key in the store.
-	// End defined as nil will iterate till the last key
-	// in the store.
+	// Providing an `End` value equal to or smaller than this value
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
 	Start []byte
-	End   []byte
 
-	// Reverse the direction of the iteration.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically smaller than this
+	// value.
+	//
+	// Providing an End value equal to or smaller than Start
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
+	End []byte
+
+	// Reverse the direction of the iteration, returning items in
+	// lexographically descending order of their keys.
 	Reverse bool
 
-	// Only iterate through keys. Calling Value() on the
-	// iterator *must* return nil (no error).
+	// Only iterate through keys. Calling Value on the
+	// iterator will return nil and no error.
+	//
+	// This option is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/34
+	//
+	// It is very likely ignored for the memory store iteration:
+	// https://github.com/sourcenetwork/corekv/issues/33
 	KeysOnly bool
```

Comment:
discussion: We could remove this option, replacing it with a keys-only iterator func, that lacks `Value` related stuff:

```
type KeyIterator interface {
   // e.g. no `Value` function
}

type Store interface {
   Iterator() Iterator
   KeyIterator() KeyIterator
}
```

Reply:
Preference to keep this as is, keeping the `Store` interface simpler (read: less methods) is a priority imo. I don't think we win much in practice for the type safety of a `KeyIterator` with no `Value` method, at the cost of maintaining a larger interface that all stores need to implement separately, vs just tracking the value in the options struct and omitting the value.

> It is very likely ignored for the memory store iteration:
It seems like the referenced bug is partly the inspiration for wanting this separation of interfaces. But I think that its trying to be too clever/complex with the type system to fix what is a relatively basic bug. Yes the bug could never exist in the first place if we had these proposed interfaces, but I don't that warrants this suggestion.

Reply:
I'm going to park this for now, as this and [this](https://github.com/sourcenetwork/corekv/pull/40#discussion_r1929385801) feel like a long-running clash of styles between myself and John, where I favour slightly more complex interfaces in return for extra type safety.  Neither this change, or the one linked, should not become notably more difficult by postponing if we later chose to make the change(s).